parboiled-core/src/main/scala/org/parboiled2/Parser.scala

```diff
@@ -449,7 +449,7 @@ abstract class Parser(initialValueStackSize: Int = 16,
    * THIS IS NOT PUBLIC API and might become hidden in future. Use only if you know what you are doing!
    */
   def __matchMap(m: Map[String, Any]): Boolean = {
-    val keys = m.keysIterator
+    val keys = m.keys.toSeq.sortBy(-_.length).iterator
```

Comment:
I think we can actually make this more efficient.
We don't need to sort the whole map, only the prefix up until the match.
Check out [this SO answer](http://stackoverflow.com/questions/12639320/lazy-sorts-of-iterators-in-scala) for an idea of how this could be done.